File: php/phplog/src/StreamDemo.php
<?php

namespace Laixhe\Phplog;

/**
 * 流/迭代器专题：PHP 数组函数与生成器（Generator）。
 * 对应 Rust rustlog/src/iterators.rs 与 Java StreamDemo。
 *
 * 对应关系速查：
 * - map          → array_map / 生成器
 * - filter       → array_filter
 * - filter_map   → array_filter + array_map / array_map 后过滤 null
 * - take(n)      → array_slice(0, n) / 生成器 break
 * - skip(n)      → array_slice(n)
 * - step_by(n)   → range(start, end, step)
 * - enumerate    → foreach 带 key / array_keys
 * - zip          → array_map 多数组 / array_combine
 * - flatten      → 循环 array_merge
 * - fold/reduce  → array_reduce
 * - partition    → 循环分组 / array_filter + array_diff
 * - collect      → array_values / iterator_to_array
 *
 * ⭐ 生成器（yield）是 PHP 的惰性迭代器，与 Rust 迭代器/Java Stream 一样
 *   在消费时才真正计算，适合处理大数据流。
 */
final class StreamDemo
{
    // ============ 基础：数组的三种遍历方式 ============
    public static function basics(): void
    {
        $v = [10, 20, 30];

        // 1) foreach 遍历（最常用）
        echo 'foreach: ';
        foreach ($v as $x) {
            echo $x, ' ';
        }
        echo PHP_EOL;

        // 2) 索引遍历
        for ($i = 0; $i < count($v); $i++) {
            echo "索引 {$i} = {$v[$i]}", PHP_EOL;
        }

        // 3) 带索引遍历（对应 enumerate）
        foreach ($v as $i => $x) {
            echo "enumerate: [{$i}] = '{$x}'", PHP_EOL;
        }

        // 生成器：手动 next()（展示迭代器底层）
        $it = (function () {
            yield 'a';
            yield 'b';
            yield 'c';
        })();
        echo '手动 next: ', $it->current(), ' '; // a
        $it->next();
        echo $it->current(), ' '; // b
        $it->next();
        echo $it->current(), ' '; // c
        $it->next();
        echo '结束? ', var_export(!$it->valid(), true), PHP_EOL; // true
    }

    // ============ 中间操作（惰性：生成器）============
    public static function adapters(): void
    {
        // ---- map：对每个元素做变换 ----
        $squares = array_map(fn ($x) => $x * $x, range(1, 10));
        echo '1..10 平方: ', implode(',', $squares), PHP_EOL;
        // 1,4,9,16,25,36,49,64,81,100

        // ---- filter：只保留满足条件的元素 ----
        $words = ['rust', 'go', 'python', 'java', 'c++', 'js'];
        $shortWords = array_values(array_filter($words, fn ($w) => strlen($w) <= 3));
        echo '长度≤3 的单词: ', implode(',', $shortWords), PHP_EOL; // go,c++,js

        // ---- filter_map 二合一：把能解析为数字的挑出来（对应 Rust filter_map）----
        $strs = ['123', 'abc', '456', 'not_a_num', '789'];
        $nums = [];
        foreach ($strs as $s) {
            $n = filter_var($s, FILTER_VALIDATE_INT);
            if ($n !== false) {
                $nums[] = $n; // 解析成功才保留
            }
        }
        echo 'filter_map 选出合法数字: ', implode(',', $nums), PHP_EOL; // 123,456,789

        // ---- take(n) / skip(n) ----
        echo 'array_slice take(3): ', implode(',', array_slice(range(1, 10), 0, 3)), PHP_EOL; // 1,2,3
        echo 'array_slice skip(7): ', implode(',', array_slice(range(1, 10), 7)), PHP_EOL;    // 8,9,10

        // ---- step_by(n)：每 n 个取一个 ----
        $stepped = range(0, 20, 5);
        echo 'step_by(5) 0..20: ', implode(',', $stepped), PHP_EOL; // 0,5,10,15,20

        // ---- zip：把两个数组一一配对 ----
        $names = ['Alice', 'Bob', 'Charlie'];
        $scores = [95, 87, 92];
        $pairs = array_map(null, $names, $scores); // array_map(null, ...) 实现 zip
        echo 'zip 配对: ', json_encode($pairs, JSON_UNESCAPED_UNICODE), PHP_EOL;
        // [["Alice",95],["Bob",87],["Charlie",92]]

        // ---- chain：把两个数组首尾相接（对应 chain）----
        $chained = array_merge(range(1, 3), range(10, 12));
        echo 'chain: ', implode(',', $chained), PHP_EOL; // 1,2,3,10,11,12

        // ---- flatten：把嵌套数组展平一层（对应 flatten / flat_map）----
        $nested = [[1, 2], [3, 4, 5], [6]];
        $flat = [];
        foreach ($nested as $sub) {
            array_push($flat, ...$sub); // 展平
        }
        echo 'flatMap 展平: ', implode(',', $flat), PHP_EOL; // 1,2,3,4,5,6

        // 把每个单词的字符展开（对应 flat_map chars）
        $chars = str_split(implode('', ['hello', 'world']));
        echo 'flat_map 展开字符: ', implode(',', $chars), PHP_EOL; // h,e,l,l,o,w,o,r,l,d
    }

    // ============ 终结操作（真正触发计算）============
    public static function consumers(): void
    {
        $v = [3, 1, 4, 1, 5, 9, 2, 6];

        // ---- collect：收集到目标数组 ----
        $doubled = array_map(fn ($x) => $x * 2, $v);
        echo 'collect 到数组: ', implode(',', $doubled), PHP_EOL;

        // ---- sum / count / min / max ----
        echo 'sum=', array_sum($v), ' count=', count($v), PHP_EOL; // sum=31 count=8
        echo 'min=', min($v), ' max=', max($v), PHP_EOL;           // min=1 max=9

        // ---- any / all：是否「有一个」/「全部」满足条件 ----
        $anyGt10 = count(array_filter($v, fn ($x) => $x > 10)) > 0;
        $allGt0 = count(array_filter($v, fn ($x) => $x <= 0)) === 0;
        echo 'any > 10? ', var_export($anyGt10, true), '  all > 0? ', var_export($allGt0, true), PHP_EOL;
        // false true

        // ---- forEach：对每个元素执行副作用 ----
        echo 'forEach: ';
        array_walk($v, fn ($x) => print($x . ' '));
        echo PHP_EOL;

        // ---- reduce：累积聚合（对应 fold / reduce）----
        $foldSum = array_reduce(range(1, 10), fn ($acc, $x) => $acc + $x, 0);
        echo 'array_reduce 累加 1..10 = ', $foldSum, PHP_EOL; // 55

        // ---- partition：按条件分成两组 ----
        $even = [];
        $odd = [];
        foreach ($v as $x) {
            if ($x % 2 === 0) {
                $even[] = $x;
            } else {
                $odd[] = $x;
            }
        }
        echo 'partition 奇偶分: 偶=[', implode(',', $even), ']  奇=[', implode(',', $odd), ']', PHP_EOL;
    }

    // ============ 综合实战：用数组函数处理复杂查询 ============
    // 场景：给定一批员工（部门，年龄，月薪），求出「R&D 部门 30 岁以上员工的平均月薪」。
    public static function practice(): void
    {
        $staff = [
            ['dept' => 'R&D', 'age' => 28, 'salary' => 30000],
            ['dept' => 'R&D', 'age' => 35, 'salary' => 45000],
            ['dept' => 'R&D', 'age' => 42, 'salary' => 60000],
            ['dept' => 'HR', 'age' => 32, 'salary' => 18000],
            ['dept' => 'R&D', 'age' => 25, 'salary' => 22000],
            ['dept' => 'Sale', 'age' => 38, 'salary' => 25000],
        ];

        // 要求：R&D 部门 + 30 岁以上 → 平均月薪
        $filtered = array_filter(
            $staff,
            fn ($e) => $e['dept'] === 'R&D' && $e['age'] >= 30 // 先筛选部门，再筛选年龄
        );
        $salaries = array_column($filtered, 'salary'); // 提取月薪
        $avg = count($salaries) === 0 ? 0 : array_sum($salaries) / count($salaries);
        printf("R&D 30+ 员工平均月薪: %.0f 元/月" . PHP_EOL, $avg); // (45000+60000)/2 = 52500
    }
}

File: java/javalog/src/main/java/com/laixhe/javalog/demo/StreamDemo.java
package com.laixhe.javalog.demo;

import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

/**
 * 迭代器/流专题：Java Stream 是「惰性求值的元素序列」，对应 Rust 迭代器。
 * 对应 Rust rustlog/src/iterators.rs。
 *
 * 两类操作：
 * - 中间操作（Adapter）：map / filter / limit / skip / flatMap 等，惰性执行，返回新流
 * - 终结操作（Consumer）：collect / sum / count / reduce / forEach 等，真正触发计算
 *
 * 对应关系速查：
 * - map → map / mapToInt；filter → filter；filter_map → filter+map / flatMap(Optional::stream)
 * - take(n) → limit(n)；skip(n) → skip(n)；step_by → IntStream.iterate
 * - enumerate → IntStream.range + 索引；zip → 两个流配对；chain → Stream.concat
 * - flatten → flatMap；fold/reduce → reduce；partition → partitioningBy
 */
public final class StreamDemo {

    private StreamDemo() {
    }

    // ============ 基础：流的三种创建方式 ============
    public static void basics() {
        // 1) 从集合创建（对应 iter()，只读借用）
        List<Integer> v = List.of(10, 20, 30);
        System.out.print("stream(): ");
        v.stream().forEach(x -> System.out.print(x + " "));
        System.out.println();

        // 2) 从数组创建
        int[] arr = {1, 2, 3};
        int sum = Arrays.stream(arr).sum();
        System.out.println("Arrays.stream 求和: " + sum); // 6

        // 3) 范围创建（对应 Range 迭代器 1..=10，包含两端）
        int total = IntStream.rangeClosed(1, 10).sum();
        System.out.println("rangeClosed(1,10) 求和: " + total); // 55

        // 手动迭代（对应手动调用 next()，展示迭代器底层）
        Iterator<Integer> it = v.iterator();
        System.out.println("手动 next: " + it.next() + " " + it.next() + " " + it.next() + " " + it.hasNext());
        // 10 20 30 false
    }

    // ============ 中间操作（惰性！）============
    public static void adapters() {
        // ---- map：对每个元素做变换 ----
        List<Integer> squares = IntStream.rangeClosed(1, 10).map(x -> x * x).boxed().toList();
        System.out.println("1..=10 平方: " + squares); // [1,4,9,16,25,36,49,64,81,100]

        // ---- filter：只保留满足条件的元素 ----
        List<String> words = List.of("rust", "go", "python", "java", "c++", "js");
        List<String> shortWords = words.stream().filter(w -> w.length() <= 3).toList();
        System.out.println("长度≤3 的单词: " + shortWords); // [go, c++, js]

        // ---- filter + map 二合一（对应 filter_map：把能解析为数字的挑出来）----
        List<String> strs = List.of("123", "abc", "456", "not_a_num", "789");
        List<Integer> nums = strs.stream()
                .map(s -> {
                    try {
                        return Optional.of(Integer.parseInt(s));
                    } catch (NumberFormatException e) {
                        return Optional.<Integer>empty();
                    }
                })
                .flatMap(Optional::stream) // 对应 filter_map
                .toList();
        System.out.println("filter_map 选出合法数字: " + nums); // [123, 456, 789]

        // ---- take(n) / skip(n)：取前 n 个 / 跳过前 n 个 ----
        System.out.println("limit(3): " + IntStream.rangeClosed(1, 10).limit(3).boxed().toList()); // [1,2,3]
        System.out.println("skip(7): " + IntStream.rangeClosed(1, 10).skip(7).boxed().toList());   // [8,9,10]

        // ---- step_by(n)：每 n 个取一个（对应 step_by(5) 0..=20）----
        List<Integer> stepped = IntStream.iterate(0, i -> i <= 20, i -> i + 5).boxed().toList();
        System.out.println("step_by(5) 0..=20: " + stepped); // [0,5,10,15,20]

        // ---- enumerate：给每个元素加上索引（对应 enumerate）----
        IntStream.range(0, "Rust".length())
                .forEach(i -> System.out.println("  enumerate: [" + i + "] = '" + "Rust".charAt(i) + "'"));
        // [0]='R' [1]='u' [2]='s' [3]='t'

        // ---- zip：把两个流的元素一一配对（长度以较短为准）----
        List<String> names = List.of("Alice", "Bob", "Charlie");
        List<Integer> scores = List.of(95, 87, 92);
        List<String> pairs = IntStream.range(0, Math.min(names.size(), scores.size()))
                .mapToObj(i -> "(" + names.get(i) + "," + scores.get(i) + ")")
                .toList();
        System.out.println("zip 配对: " + pairs); // [(Alice,95), (Bob,87), (Charlie,92)]

        // ---- chain：把两个流首尾相接（对应 chain）----
        List<Integer> chained = IntStream.concat(
                IntStream.rangeClosed(1, 3),
                IntStream.rangeClosed(10, 12)).boxed().toList();
        System.out.println("chain: " + chained); // [1,2,3,10,11,12]

        // ---- flatMap：把嵌套的流展平一层（对应 flatten / flat_map）----
        List<List<Integer>> nested = List.of(List.of(1, 2), List.of(3, 4, 5), List.of(6));
        List<Integer> flat = nested.stream().flatMap(List::stream).toList();
        System.out.println("flatMap 展平: " + flat); // [1,2,3,4,5,6]

        // 把每个单词的字符展开（对应 flat_map chars）
        List<Character> chars = List.of("hello", "world").stream()
                .flatMap(w -> w.chars().mapToObj(c -> (char) c))
                .toList();
        System.out.println("flat_map 展开字符: " + chars); // [h,e,l,l,o,w,o,r,l,d]
    }

    // ============ 终结操作（真正触发计算）============
    public static void consumers() {
        List<Integer> v = List.of(3, 1, 4, 1, 5, 9, 2, 6);

        // ---- collect：收集到目标集合（对应 collect）----
        List<Integer> doubled = v.stream().map(x -> x * 2).toList();
        System.out.println("collect 到 List: " + doubled);

        Set<Integer> unique = v.stream().collect(Collectors.toSet());
        System.out.println("collect 到 Set（去重）: " + unique);

        // ---- sum / count ----
        int sum = v.stream().mapToInt(Integer::intValue).sum();
        long count = v.stream().count();
        System.out.println("sum=" + sum + ", count=" + count); // sum=31, count=8

        // ---- min / max（返回 Optional，空流是 empty）----
        System.out.println("min=" + v.stream().min(Integer::compareTo)
                + " max=" + v.stream().max(Integer::compareTo)); // Optional[1] Optional[9]

        // ---- any / all：是否「有一个」/「全部」满足条件 ----
        System.out.println("any > 10? " + v.stream().anyMatch(x -> x > 10)); // false
        System.out.println("all > 0?  " + v.stream().allMatch(x -> x > 0));  // true

        // ---- forEach：对每个元素执行副作用（对应 for_each）----
        System.out.print("forEach: ");
        IntStream.rangeClosed(1, 5).forEach(x -> System.out.print(x + " "));
        System.out.println();

        // ---- reduce / fold：累积聚合（对应 fold / reduce）----
        int foldSum = IntStream.rangeClosed(1, 10).reduce(0, Integer::sum); // 带初始值
        System.out.println("reduce 累加 1..=10 = " + foldSum); // 55

        Optional<Integer> maxReduce = v.stream().reduce(Integer::max);
        System.out.println("reduce 手动求最大值 = " + maxReduce); // Optional[9]

        // ---- partition：按条件分成两组（对应 partition）----
        Map<Boolean, List<Integer>> partition = v.stream()
                .collect(Collectors.partitioningBy(x -> x % 2 == 0));
        System.out.println("partition 奇偶分: 偶=" + partition.get(true) + "  奇=" + partition.get(false));
    }

    // ============ 综合实战：用流处理复杂查询 ============
    // 场景：给定一批员工（部门，年龄，月薪），求出「R&D 部门 30 岁以上员工的平均月薪」。
    public static void practice() {
        record Employee(String dept, int age, int salary) {
        }

        List<Employee> staff = List.of(
                new Employee("R&D", 28, 30000),
                new Employee("R&D", 35, 45000),
                new Employee("R&D", 42, 60000),
                new Employee("HR", 32, 18000),
                new Employee("R&D", 25, 22000),
                new Employee("Sale", 38, 25000));

        // 要求：R&D 部门 + 30 岁以上 → 平均月薪
        var query = staff.stream()
                .filter(e -> e.dept().equals("R&D"))   // 先筛选部门
                .filter(e -> e.age() >= 30)            // 再筛选年龄
                .mapToInt(Employee::salary)            // 提取月薪
                .toArray();

        double avg = query.length == 0 ? 0.0 : Arrays.stream(query).average().orElse(0);
        System.out.printf("R&D 30+ 员工平均月薪: %.0f 元/月%n", avg); // (45000+60000)/2 = 52500
    }
}
